File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/VariableEditorActivity.java
package ch.rmy.android.http_shortcuts;

import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.EditText;

import com.satsuware.usefulviews.LabelledSpinner;

import butterknife.Bind;
import ch.rmy.android.http_shortcuts.realm.Controller;
import ch.rmy.android.http_shortcuts.realm.models.Variable;
import ch.rmy.android.http_shortcuts.utils.ArrayUtil;
import ch.rmy.android.http_shortcuts.utils.GsonUtil;
import ch.rmy.android.http_shortcuts.utils.SimpleTextWatcher;
import ch.rmy.android.http_shortcuts.utils.ViewUtil;
import ch.rmy.android.http_shortcuts.variables.Variables;

public class VariableEditorActivity extends BaseActivity {

    public final static String EXTRA_VARIABLE_ID = "ch.rmy.android.http_shortcuts.VariableEditorActivity.variable_id";
    private static final String STATE_JSON_VARIABLE = "variable_json";

    @Bind(R.id.input_variable_type)
    LabelledSpinner typeSpinner;
    @Bind(R.id.input_variable_name)
    EditText nameView;
    @Bind(R.id.input_variable_title)
    EditText titleView;

    private Controller controller;
    private Variable oldVariable;
    private Variable variable;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_variable_editor);

        controller = destroyer.own(new Controller(this));

        long variableId = getIntent().getLongExtra(EXTRA_VARIABLE_ID, 0);
        if (savedInstanceState != null && savedInstanceState.containsKey(STATE_JSON_VARIABLE)) {
            variable = GsonUtil.fromJson(savedInstanceState.getString(STATE_JSON_VARIABLE), Variable.class);
        } else {
            variable = variableId == 0 ? Variable.createNew() : controller.getDetachedVariableById(variableId);
        }
        if (variable == null) {
            finish();
            return;
        }
        oldVariable = variableId == 0 ? Variable.createNew() : controller.getDetachedVariableById(variableId);

        initViews();
    }

    private void initViews() {
        nameView.setText(variable.getKey());
        titleView.setText(variable.getTitle());
        final ColorStateList defaultColor = nameView.getTextColors();
        nameView.addTextChangedListener(new SimpleTextWatcher() {
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (Variables.isValidVariableName(s.toString())) {
                    nameView.setTextColor(defaultColor);
                    nameView.setError(null);
                } else {
                    nameView.setTextColor(Color.RED);
                    nameView.setError(s.length() == 0 ? null : getString(R.string.warning_invalid_variable_name));
                }
            }
        });

        typeSpinner.setItemsArray(Variable.getTypeOptions(this));
        ViewUtil.hideErrorLabel(typeSpinner);
        typeSpinner.setSelection(ArrayUtil.findIndex(Variable.TYPE_OPTIONS, variable.getType()));

        setTitle(variable.isNew() ? R.string.create_variable : R.string.edit_variable);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.variable_editor_activity_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    protected int getNavigateUpIcon() {
        return R.drawable.ic_clear;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home: {
                confirmClose();
                return true;
            }
            case R.id.action_save_variable: {
                trySave();
                return true;
            }
        }
        return super.onOptionsItemSelected(item);
    }

    private void confirmClose() {
        //TODO: Check for changes and prompt
        finish();
    }

    private void trySave() {
        compileVariable();
        if (validate()) {
            controller.persist(variable);
            finish();
        }
    }

    private void compileVariable() {
        variable.setTitle(titleView.getText().toString().trim());
        variable.setKey(nameView.getText().toString().trim());
        variable.setType(Variable.TYPE_OPTIONS[typeSpinner.getSpinner().getSelectedItemPosition()]);
    }

    private boolean validate() {
        //TODO: Validate inputs and display error messages
        //TODO: Check if the variable key is unique
        return true;
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        compileVariable();
        outState.putString(STATE_JSON_VARIABLE, GsonUtil.toJson(variable));
    }

}

File: HTTPShortcuts/app/src/main/kotlin/ch/rmy/android/http_shortcuts/activities/variables/VariableEditorActivity.kt
package ch.rmy.android.http_shortcuts.activities.variables

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import ch.rmy.android.http_shortcuts.R
import ch.rmy.android.http_shortcuts.activities.BaseActivity
import ch.rmy.android.http_shortcuts.data.models.Variable
import ch.rmy.android.http_shortcuts.databinding.ActivityVariableEditorBinding
import ch.rmy.android.http_shortcuts.dialogs.DialogBuilder
import ch.rmy.android.http_shortcuts.extensions.attachTo
import ch.rmy.android.http_shortcuts.extensions.bindViewModel
import ch.rmy.android.http_shortcuts.extensions.consume
import ch.rmy.android.http_shortcuts.extensions.focus
import ch.rmy.android.http_shortcuts.extensions.observeTextChanges
import ch.rmy.android.http_shortcuts.extensions.visible
import ch.rmy.android.http_shortcuts.utils.BaseIntentBuilder
import ch.rmy.android.http_shortcuts.variables.Variables
import ch.rmy.android.http_shortcuts.variables.types.HasTitle
import ch.rmy.android.http_shortcuts.variables.types.VariableEditorFragment
import ch.rmy.android.http_shortcuts.variables.types.VariableTypeFactory

class VariableEditorActivity : BaseActivity() {

    private val variableId: String? by lazy {
        intent.getStringExtra(EXTRA_VARIABLE_ID)
    }
    private val preferredType: String? by lazy {
        intent.getStringExtra(EXTRA_VARIABLE_TYPE)
    }

    private val viewModel: VariableEditorViewModel by bindViewModel()

    private val variable by lazy {
        viewModel.getVariable()
    }

    private lateinit var binding: ActivityVariableEditorBinding

    private var fragment: VariableEditorFragment<*>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = applyBinding(ActivityVariableEditorBinding.inflate(layoutInflater))
        initViewModel()
        initViews()
    }

    private fun initViewModel() {
        viewModel.variableType = preferredType ?: Variable.TYPE_CONSTANT
        viewModel.variableId = variableId
    }

    private fun initViews() {
        setSubtitle(VariableTypes.getTypeName(variable.type))
        binding.inputVariableKey.setText(variable.key)
        binding.inputVariableTitle.setText(variable.title)
        val defaultColor = binding.inputVariableKey.textColors
        binding.inputVariableKey
            .observeTextChanges()
            .subscribe { text ->
                if (text.isEmpty() || Variables.isValidVariableKey(text.toString())) {
                    binding.inputVariableKey.setTextColor(defaultColor)
                    binding.inputVariableKey.error = null
                } else {
                    binding.inputVariableKey.setTextColor(Color.RED)
                    binding.inputVariableKey.error = getString(R.string.warning_invalid_variable_key)
                }
            }
            .attachTo(destroyer)

        binding.inputUrlEncode.isChecked = variable.urlEncode
        binding.inputJsonEncode.isChecked = variable.jsonEncode
        binding.inputAllowShare.isChecked = variable.isShareText

        setTitle(if (variable.isNew) R.string.create_variable else R.string.edit_variable)

        updateTypeEditor()
    }

    private fun updateTypeEditor() {
        compileVariable()
        val variableType = VariableTypeFactory.getType(variable.type)

        fragment = variableType.getEditorFragment(supportFragmentManager)

        binding.dialogTitleContainer.visible = variableType is HasTitle

        fragment?.let { fragment ->
            supportFragmentManager
                .beginTransaction()
                .replace(R.id.variable_type_fragment_container, fragment, variableType.tag)
                .commitAllowingStateLoss()
        }
    }

    fun onFragmentStarted() {
        fragment?.updateViews(variable)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.variable_editor_activity_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override val navigateUpIcon = R.drawable.ic_clear

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        android.R.id.home -> consume { confirmClose() }
        R.id.action_save_variable -> consume { trySave() }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        confirmClose()
    }

    private fun confirmClose() {
        compileVariable()
        if (viewModel.hasChanges()) {
            DialogBuilder(context)
                .message(R.string.confirm_discard_changes_message)
                .positive(R.string.dialog_discard) { finish() }
                .negative(R.string.dialog_cancel)
                .showIfPossible()
        } else {
            finish()
        }
    }

    private fun trySave() {
        compileVariable()
        if (validate()) {
            viewModel.save()
                .subscribe {
                    finish()
                }
                .attachTo(destroyer)
        }
    }

    private fun compileVariable() {
        fragment?.compileIntoVariable(variable)
        variable.title = binding.inputVariableTitle.text.toString().trim { it <= ' ' }
        variable.key = binding.inputVariableKey.text.toString()
        variable.urlEncode = binding.inputUrlEncode.isChecked
        variable.jsonEncode = binding.inputJsonEncode.isChecked
        variable.isShareText = binding.inputAllowShare.isChecked
    }

    private fun validate(): Boolean {
        if (variable.key.isEmpty()) {
            binding.inputVariableKey.error = getString(R.string.validation_key_non_empty)
            binding.inputVariableKey.focus()
            return false
        }
        if (viewModel.isKeyAlreadyInUsed()) {
            binding.inputVariableKey.error = getString(R.string.validation_key_already_exists)
            binding.inputVariableKey.focus()
            return false
        }
        return fragment == null || fragment!!.validate()
    }

    override fun onStop() {
        super.onStop()
        compileVariable()
    }

    class IntentBuilder(context: Context) : BaseIntentBuilder(context, VariableEditorActivity::class.java) {

        fun variableType(type: String) = also {
            intent.putExtra(EXTRA_VARIABLE_TYPE, type)
        }

        fun variableId(variableId: String) = also {
            intent.putExtra(EXTRA_VARIABLE_ID, variableId)
        }

    }

    companion object {

        private const val EXTRA_VARIABLE_ID = "ch.rmy.android.http_shortcuts.activities.variables.VariableEditorActivity.variable_id"
        private const val EXTRA_VARIABLE_TYPE = "ch.rmy.android.http_shortcuts.activities.variables.VariableEditorActivity.variable_type"

    }

}
